driver/embedding_driver.hpp

```diff
@@ -0,0 +1,434 @@
+            continue;
+
+        if(embedding_idx >= 0 && embedding_idx < num_embeddings)
+        {
+            Tcheck scale =
+                indices_freq
+                    ? (static_cast<Tcheck>(1.0f) / static_cast<Tcheck>(indices_freq[input_idx]))
+                    : static_cast<Tcheck>(1.0f);
+            size_t weight_grad_idx = weightGrad_tv.get_tensor_view_idx({embedding_idx, j});
+            tensor_layout_t<4> outGrad_layout(outGrad_tv, o);
+            weightGradHost[weight_grad_idx] +=
+                outputGrad[outGrad_tv.get_tensor_view_idx(outGrad_layout)] * scale;
+        }
+    }
+
+    return 0;
+}
+
+template <typename Tgpu, typename Tref>
+class EmbeddingDriver : public Driver
+{
+public:
+    EmbeddingDriver() : Driver()
+    {
+        miopenCreateTensorDescriptor(&inputTensor);
+        miopenCreateTensorDescriptor(&outputTensorGrad);
+        miopenCreateTensorDescriptor(&weightTensorGrad);
+
+        data_type = miopen_type<Tgpu>{};
+    }
+
+    std::vector<int> ComputeStrides(std::vector<int> input);
+    int AddCmdLineArgs() override;
+    int ParseCmdLineArgs(int argc, char* argv[]) override;
+    InputFlags& GetInputFlags() override { return inflags; }
+
+    int GetandSetData() override;
+    std::vector<int> GetInputTensorLengthsFromCmdLine();
+
+    int AllocateBuffersAndCopy() override;
+
+    int RunForwardGPU() override;
+    int RunForwardCPU(); // Verify implements it
+
+    int RunBackwardGPU() override;
+    int RunBackwardCPU(); // Verify implements it
+
+    Tref GetTolerance();
+
+    int VerifyBackward() override;
+    int VerifyForward() override;
+    ~EmbeddingDriver() override
+    {
+        miopenDestroyTensorDescriptor(inputTensor);
+        miopenDestroyTensorDescriptor(outputTensorGrad);
+        miopenDestroyTensorDescriptor(weightTensorGrad);
+    }
+
+private:
+    InputFlags inflags;
+
+    int forw;
+    bool isContiguous;
+
+    // Backwards
+    miopenTensorDescriptor_t inputTensor;
+    miopenTensorDescriptor_t outputTensorGrad;
+    miopenTensorDescriptor_t weightTensorGrad;
+
+    std::unique_ptr<GPUMem> in_dev;
+    std::unique_ptr<GPUMem> outGrad_dev;
+    std::unique_ptr<GPUMem> weightGrad_dev;
+
+    std::vector<int64_t> in;
+    std::vector<Tgpu> outGrad;
+    std::vector<Tgpu> weightGrad;
+    std::vector<Tref> weightGradHost;
+
+    std::vector<int32_t> indices_freq;
+    std::unique_ptr<GPUMem> indices_freq_dev;
+
+    int64_t padding_idx;
+    bool scale_grad_by_freq;
+};
+
+template <typename Tgpu, typename Tref>
+int EmbeddingDriver<Tgpu, Tref>::ParseCmdLineArgs(int argc, char* argv[])
+{
+    inflags.Parse(argc, argv);
+
+    if(inflags.GetValueInt("time") == 1)
+    {
+        miopenEnableProfiling(GetHandle(), true);
+    }
+
+    forw = inflags.GetValueInt("forw");
+
+    if(forw != 2)
+    {
+        MIOPEN_THROW("Invalid Forward|Backward Mode");
+    }
+
+    isContiguous       = inflags.GetValueInt("is_contiguous") == 1 ? true : false;
```

Comment:
C++ will perform implicitly conversion from integer `1` to the target type of your variable 
```suggestion
    isContiguous       = inflags.GetValueInt("is_contiguous")
```

Reply:
btw, I also think it's not logical to use `== 1 ? true : false;`. It should be `== 0 ? false : true`, because semantically, only numerical `0` value is treated as `false`, any other numerical values are all `true`

Reply:
updated.